File: app/src/main/java/at/shockbytes/dante/network/google/GoogleBooksDownloader.java
package at.shockbytes.dante.network.google;

import javax.inject.Inject;

import at.shockbytes.dante.network.BookDownloader;
import at.shockbytes.dante.util.books.BookSuggestion;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;


/**
 * @author Martin Macheiner
 *         Date: 13.02.2017.
 */

public class GoogleBooksDownloader implements BookDownloader {

    private GoogleBooksApi api;

    @Inject
    public GoogleBooksDownloader(GoogleBooksApi api) {
        this.api = api;
    }

    @Override
    public Observable<BookSuggestion> downloadBookSuggestion(String isbn) {
        return api.downloadBookSuggestion(isbn)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io());
    }
}

File: app/src/main/java/at/shockbytes/dante/network/google/GoogleBooksDownloader.kt
package at.shockbytes.dante.network.google

import at.shockbytes.dante.book.BookSuggestion
import at.shockbytes.dante.network.BookDownloader
import at.shockbytes.dante.util.scheduler.SchedulerFacade
import io.reactivex.Observable

/**
 * Author:  Martin Macheiner
 * Date:    13.02.2017
 */
class GoogleBooksDownloader(
    private val api: GoogleBooksApi,
    private val schedulerFacade: SchedulerFacade
) : BookDownloader {

    override fun downloadBook(isbn: String): Observable<BookSuggestion> {
        return api
            .downloadBookSuggestion(isbn)
            .observeOn(schedulerFacade.ui)
            .subscribeOn(schedulerFacade.io)
    }
}
